src/crypto/slow-hash.c

```diff
@@ -47,6 +49,24 @@
 extern int aesb_single_round(const uint8_t *in, uint8_t*out, const uint8_t *expandedKey);
 extern int aesb_pseudo_round(const uint8_t *in, uint8_t *out, const uint8_t *expandedKey);
 
+#define VARIANT1_1(p) \
+  do if (variant > 0) \
+  { \
+    uint8_t tmp = ((const uint8_t*)p)[11]; \
+    uint8_t tmp1 = (tmp>>4)&1, tmp2 = (tmp>>5)&1, tmp3 = tmp1^tmp2; \
+    uint8_t tmp0 = nonce_flag ? tmp3 : tmp1 + 1; \
+    ((uint8_t*)p)[11] = (tmp & 0xef) | (tmp0<<4); \
```

Comment:
I think your logic table and the resulting code is incorrect. For example, 01 / 0 should result in p[11] (bits 5, 4) = 10, not 11. I tested mine by running through all 256 values and comparing results with the original macro.

Reply:
I did get my masks wrong, in that they were masking one bit over from this patch (the bitshifting is zero based, dummy me), but otherwise the table is correct AFAIK. The 01 / 0  case you mentioned:

tmp = 0y00010000
tmp1 = 0y00000001
tmp2 = 0y00000000
tmp3 = 0y00000001
tmp0 = 0y00000010
p[11] = 0y00110000

The difference in our observations are that I was referring to (5,4) if bits indexes are (7,0). The code I wrote probably confused things because the constants needed to be 0x30, 0x20, and 0x10 (i.e. I made the same mixup when writing the constants that you likely did when analyzing the values).

P.S. -> yes I was suggesting a variant to remove the bias in the output. This probably doesn't matter much because the remainder of the algorithm will "whiten" this bias anyway. The primary reason to reject my change is that it might be easier to modify an existing ASIC. I think it would be the same - with either change a hardware designer might consider the equivalent of cmovs with constants instead of arithmetic operations. Although this getting a bit far out of my area of expertise (specifically the difficulty/expense at designing the types of circuits I am referencing).

P.P.S. -> No.